File: src/Belajar_SOLID_11_Open_Close_Principle/Java/Main.java
package Belajar_SOLID_11_Open_Close_Principle.Java;

class Product {
    // Product Entitas
}

class ShippingOrderService{
    public void chekout(Product product, Shipping shipping){
        int costShipping = shipping.calculate(product);
    }
}

abstract class Shipping{
    abstract public int calculate(Product product);
}

class JNEShipping extends Shipping{
    @Override
    public int calculate(Product product) {
        return 0;
    }
}

class TIKIShipping extends Shipping{
    @Override
    public int calculate(Product product) {
        return 0;
    }
}

// Penambahan class baru tanpa mengubah kode di ShippingOrederService
class SiCepatShipping extends Shipping{
    @Override
    public int calculate(Product product) {
        return 0;
    }
}
//    Ini bisa dibuktikan ketika ingin menambahkan
//    jenis pengiriman baru, kita cukup membuat class baru tanpa
//    harus mengubah kode yang berada di dalam class ShippingOrderService.

File: src/Belajar_SOLID_11_Open_Close_Principle/Kotlin/Main.kt
package Belajar_SOLID_11_Open_Close_Principle.Kotlin

class Product{
    //    Product Entitas
}

class ShippingOrderService{
    fun checkout(product: Product, shipping: Shipping){
        val costShipping = shipping.calculateProduct(product)
        /* code to do check */
    }
}

abstract class Shipping{
    abstract fun calculateProduct(product: Product):Int
}

class JNEShipping:Shipping(){
    override fun calculateProduct(product: Product): Int {
        TODO("Not yet implemented")
        // return  calculate amount of this type with product
    }
}

class TIKIShipping:Shipping(){
    override fun calculateProduct(product: Product): Int {
        TODO("Not yet implemented")
        // return  calculate amount of this type with product
    }
}

// Penambahan class baru tanpa mengubah kode di ShippingOrederService
class POSINDOShipping:Shipping(){
    override fun calculateProduct(product: Product): Int {
        TODO("Not yet implemented")
        // return  calculate amount of this type with product
    }
}

// Saat menerapkan open/close principle ke dalam project, kita bisa membatasi kebutuhan untuk mengubah
// kode yang telah ditulis, diuji dan di-debug. Tujuannya untuk menghindari resiko atau
// kelemahan sistem yang bisa saja terjadi. Selain itu, kita bisa menghindari ketergantungan
// dan meningkatkan fleksibilitas sistem. Tentunya ini akan meringankan proses skalabilitas
// dari sisi pengembangan perangkat lunak.
